### DISCUSSION TITLE
Is there a way to get something similar to a brewfile?

### DISCUSSION THREAD

**Discussion by mfts** [pkgxdev/pkgx#361] [Help]
Asked by `DitchComfort` via Discord

I know tea isn’t exactly brew, but is there any way to get something similar to what brewfile does? To have a own file with everything you usually have installed/downloaded appending/deleting automatically…

**Comment by jhheider** [comment]
I can think of a couple of ways to do it. You shouldn't need to, unless you're air-gapped, since tea fetches as needed, but you could easily make a script that depends on everything you want and just calls `true`.

**Comment by mfts** [comment]
This is just out there. But since `tea` injects packages in your PATH, you could create a README.md file in your HOME directory with all packages you require in a dependencies table. 

So every time you `cd $HOME` or navigate within the home directory you're packages will be available in the path. 

Now that's probably overkill and not tested but theoretically it should work.

**Reply by mxcl** [reply]
We deliberately stop descending at `$HOME` (and obv. `/`). This is following the convention of tools like `git`.

**Comment by mxcl** [comment]
I'm with Jacob in that you don’t need this. Our magic makes a lot of other scenarios “just work”.

However, we have plans for sync of sorts in the future.

I the meantime though what would be useful here is a concrete use case that needs something like a `Brewfile`. All the ones I can think of tea already deals with out the box.

**Comment by magnusviri** [comment]
Here's a use case. My university has several supercomputer clusters that researchers use. 2 of them are dedicated to researching protected health data. There are many users who are not admins. I'm not 100% sure, but I believe the clusters are firewalled from the outside world so users might be unable to install things. I can ask the administrators what happens if someone wants to install something but I'm guessing tea won't work unless the admins pre-install the software.

**Reply by jhheider** [reply]
```sh
#!/usr/bin/env tea
# baseline.sh

_="
---
args: [sh]
dependencies:
  llvm.org: ^14
  gnu.org/binutils: '*'
  stedolan.github.io/jq: '*'
[...]
---
"

true
```

`sh <(curl tea.xyz) http://intranet/www/bashline.sh`

Maybe? Requires testing and tweaking.

**Reply by magnusviri** [reply]
I can think of more use cases, mostly on centrally managed machines with multiple users who don't have admin permissions. Very rarely the machines don't have internet or all behind outgoing firewalls, but it's not often. Admins also frown on putting things in home folders, which is a big problem on managed multi-user computers because then all the home folders have to be iterated over, which is possible but a pain. This is why I'm always putting tea in /opt/tea and my daily driver is a non-admin account. I do this to always test how everything works since it's not the norm. I see more errors doing it this way. But to put things in perspective, tea is already a better admin deployment tool than brew, which requires hacking the installer script to get it on the machine (because it blocks installing as root).

Either way, as an admin, it's fairly easy to pre-install everything with tea. I just run the command and specify some form of --version or --help. If I were to deploy something like wumpus, which goes straight into interactive mode and can't be avoided with arguments, that would be a problem to deploy. This baseline.sh looks like it would work for that (not that I'd install wumpus, it's just an example--but who knows maybe some really geeky college kid out there wants to get lost in computer history).

**Comment by magnusviri** [comment]
I thought of another use case. The main benefit of Krita Linux is that it includes a lot of pre-installed software. It would be really cool if it could be reduced to a shell script that can run on any platform (including macOS) that is executed with tea to install everything and then configure it all. I'm not entirely sure if this is possible because it might require a custom kernel. I don't know. But it would be cool to have something like it.

Similarly, Linux installers come in minimal, server, and desktop versions. If a user installs the minimal, there is usually a package they can install to install everything required to turn it into a server or desktop. It would be awesome if there were a script that used tea to upgrade a minimal Linux to a desktop. Even better, it would be cool if it allowed the user to specify the desktop environment (KDE, Gnome, etc). This would make it easier to create distributions since it no longer requires a full installer, just install the favored base (e.g. Debian) and then specify the "distribution" script, which is executed by tea.

Similarly, a mail server installer script could use tea to install postfix, dovecot, SpamAssassin, and everything else needed, and then the rest of the script configured them all. This would actually make macOS a viable server OS again. The macOS Server application used to do this but doesn't anymore. Anyone who wants to use macOS as a server has to download all the sources and build them all from scratch. It was easier for me to just switch to Linux (which is why I'm also a Linux user today).

Side note, maybe brew can install server stuff, but the very fact that brew can't be run as root just kind of makes it seem like the wrong tool to be installing server software. I know the root restriction is there as a protection against doing something stupid, but maybe it says to administrators that brew isn't safe and shouldn't be trusted.

This is totally off-topic, but there are a lot of administrator tools like Ansible, Puppet, and Chef that install packages. Ansible comes with modules that run the package managers. There's a [generic module](https://docs.ansible.com/ansible/latest/collections/ansible/builtin/package_module.html) and there are modules for [apt](https://docs.ansible.com/ansible/latest/collections/ansible/builtin/apt_module.html), [yum](https://docs.ansible.com/ansible/latest/collections/ansible/builtin/yum_module.html)/[dnf](https://docs.ansible.com/ansible/latest/collections/ansible/builtin/dnf_module.html), [pacman](https://docs.ansible.com/ansible/latest/collections/community/general/pacman_module.html), and even [homebrew](https://docs.ansible.com/ansible/latest/collections/community/general/homebrew_module.html). I know it's out of scope for you guys, but I would keep in mind that someone will eventually want to add support to these admin tools for tea.

**Comment by MaxWinterstein** [comment]
Just throwing in my 2 cents, I would love to have some generic input file for tea, e.g. like pythons `requirements.txt`.

That way dependency update tools like `Renovate` or `Dependabot` might be able to integrate tea as a manager and provide updates for people who can't live on the edge and need reproducible installations.

**Reply by mxcl** [reply]
`tea.yaml` is supported. See [docs](https://docs.tea.xyz/features/developer-environments)

**Comment by mxcl** [comment]
tea magically makes things available, so I'd say brewfile is not really in our desired feature set. However adding a tea-cmd to do this is easy (add a pkg that provides tea-bundle to the pantry). So feel free.

Also we do intend however for the GUI to provide sync.

**Reply by magnusviri** [reply]
I'm still not sure what a tea-cmd is. Do you mean something like this?

pantry/projects/example.com/tea-bundle/package.yml

And the content has something like this?

```
provides:
  - bin/tea-bundle
```

**Reply by mxcl** [reply]
that's right
